azurelinuxagent/common/utils/shellutil.py

```diff
@@ -117,30 +117,51 @@ def _encode_command_output(output):
     return ustr(output, encoding='utf-8', errors="backslashreplace")
 
 
-def run_command(command):
+class CommandError(Exception):
     """
-    Wrapper for subprocess.Popen. Executes the given command and returns its stdout.
-    Logs any errors executing the command and raises an exception.
+    Exception raised by run_command when the command returns an error
     """
-    retcode = 0
+    @staticmethod
+    def _get_message(command, returncode):
+        command_name = command[0] if isinstance(command, list) and len(command) > 0 else command
+        return "'{0}' failed: {1}".format(command_name, returncode)
 
+    def __init__(self, command, returncode, stdout, stderr):
+        super(Exception, self).__init__(CommandError._get_message(command, returncode))
+        self.command = command
+        self.returncode = returncode
+        self.stdout = stdout
+        self.stderr = stderr
+
+
+def run_command(command, log_error=False):
+    """
+    Executes the given command and returns its stdout as a string.
+    If there are any errors executing the command it logs details about the failure and raises a RunCommandException;
+    if 'log_error' is True, it also logs details about the error.
+    """
     try:
         process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=False)
         stdout, stderr = process.communicate()
-        retcode = process.returncode
+        returncode = process.returncode
     except Exception as e:
-        logger.error(u"Cannot execute [{0}]. Error: [{1}]".format(command, ustr(e)))
+        if log_error:
+            logger.error(u"Command [{0}] raised unexpected exception: [{1}]", command, ustr(e))
```

Comment:
Since `command` is an array, should we convert it to a string before logging?

Reply:
sure, i'll do that